### ISSUE TITLE
Please use `geo:` URL in "Share my location"

### ISSUE THREAD

**Issue by jech** [organicmaps/organicmaps#7916]
⚠ Have you searched for similar, already existing issues?

**Describe the issue**
Please write a clear and concise description of the issue here.

When I choose "Share my location", the message proposed by OM contains a `om:` URL and a `https://omaps.org` URL, both of which are proprietary URLs intimately linked to Organic Maps.  There appears to be no way to configure OM to use a `geo:` URL instead, which works with any maps software.

**Steps to reproduce**
1. Go to ≡
2. Click on *Share my location*
3. Notice that there is no provider-neutral URL.

**Expected behaviour**

OM should share my location as a `geo:` URL, as defined in [RFC 5870](https://www.rfc-editor.org/rfc/rfc5870.html).  It should also refrain from including promotional messages ("Don't have?  Download here!")

**Comment by biodranik** [comment]
Default `geo:` URL goes without slashes (geo://) and is not decoded by any software automatically, only if it is a part of a clickable hyperlink. Do you have examples of other apps that successfully share non-hyperlinked (plain text) geo links that are properly clickable/opened by other apps?

**Comment by jech** [comment]
> Do you have examples of other apps that successfully share non-hyperlinked (plain text) geo links

[FairEmail](https://email.faircode.eu/) does the right thing.  I'll be glad to report a bug with them if you implement `geo:` in  OM's sharing dialog.

**Comment by biodranik** [comment]
Is it the only app that can decode `geo:` links? Generate links (present in every shared message) that are only useful in a specific app doesn't sound right.

**Comment by biodranik** [comment]
Does FairMail highlight to you this text and make it clickable? geo:37.786971,-122.399677?z=17

**Comment by jech** [comment]
> Does FairMail highlight to you this text and make it clickable?
> geo:37.786971,-122.399677?z=17

No, it does not.  On the other hand, it parses the following:

  geo:37,122

It looks like neither negative numbers nor decimal points are parsed
correctly.  I'll file a bug with FairEmail.

**Comment by jech** [comment]
> Is it the only app that can decode geo: links?

I don't know, FairEmail and SMS are the only messaging apps I use on my
phone.  (To be frank, I'm quite saddened by what you just demonstrated to
me, I'd expect geo: links to be more widely supported.)

> Generate links (present in every shared message) that are only useful in
> a specific app doesn't sound right.

I understand, but I think it's a chicken and egg problem.  Perhaps OM
could include geo: links in addition to the proprietary links?

**Comment by biodranik** [comment]
One of the ideas is to pass plain text coordinates instead of custom encoding: https://github.com/organicmaps/organicmaps/issues/7482

**Comment by jech** [comment]
It's a slight improvement, but it still means you need Internet access.  The promise (sadly unrealised, as you've shown to me) of geo: links is that they are application-agnostic while not relying on an Internet service.

**Comment by biodranik** [comment]
No, you don't need the internet access to open omaps.app links if you have Organic Maps installed. Click on such links should open OM. If it doesn't, please let us know so we can check and fix it.

https://omaps.app/o4B4pYZsRs/Zoo_Zürich

**Comment by jech** [comment]
You're assuming that my correspondents use OrganicMaps. Sadly, that's not 
the case.

**Comment by Menelmacar** [comment]
> Default `geo:` URL goes without slashes (geo://) and is not decoded by any software automatically, only if it is a part of a clickable hyperlink. Do you have examples of other apps that successfully share non-hyperlinked (plain text) geo links that are properly clickable/opened by other apps?

My IM app conversations (and all its forks use it too) osmand uses it too. 
Then there is my public transport android app "öffi" and basically every open street map editor /contributor app

**Comment by biodranik** [comment]
Can you please send screenshots of how these links without slashes are highlighted/clickable in these apps, or describe the steps to reproduce it?

**Comment by licaon-kter** [comment]
Isn't there a privacy issue too since everyone needs to connect to the dev site?

Eg. Share a link in a group, suddenly 10 IPs, 10 devices, 10 Android/mac/Linux/Windows devices connect to your servers.

**Comment by Menelmacar** [comment]
Conversations 
https://codeberg.org/iNPUTmice/Conversations shows a button these days, you can see what was actually send in the input field. 
![Imagepipe_2](https://github.com/organicmaps/organicmaps/assets/11552677/847f69ee-64cb-48e2-8ce9-68f68d2a73da)

The other mentioned apps open the respective location, if you share the geo string to them via the android share system. For example osmand:
![Imagepipe_0](https://github.com/organicmaps/organicmaps/assets/11552677/1c863d30-3a3b-48c3-a7f5-1a33b0c5f2be)

https://gitlab.com/oeffi/oeffi will show the nearest address from the coordinates as search target if the geo: string is shared to the app.

**Comment by biodranik** [comment]
The text with geo: is not highlighted and is not clickable.

**Comment by jech** [comment]
> The text with geo: is not highlighted and is not clickable.

Then we'll file bugs against the applications that don't handle the links correctly, as I have already done against FairEmail.  But it's a chicken and egg problem: we cannot reasonably file bugs until OM (or some other widely respected maps app) can generate them.

**Comment by Menelmacar** [comment]
Yes, it is even translated to a button. In an old version it looked like a link, highlighting it by an underscore.
The other apps are receiver or these links, so they don't even show them, they show the result.
The issue for organic maps is to create such links. The share menu will take care of the rest.

**Comment by licaon-kter** [comment]
@biodranik which of the 3 pictures are you talking about?

**Comment by biodranik** [comment]
This is how any geo:12.3456,12.3456 link will look in any app when shared:
<img width="847" alt="выява" src="https://github.com/organicmaps/organicmaps/assets/170263/dbf77435-c665-474a-9349-2161c49c6e03">

It is not clickable, so there's not much sense to share it in exactly this way.

**Comment by Menelmacar** [comment]
> This is how any geo:12.3456,12.3456 link will look in any app when shared: <img alt="выява" width="847" src="https://private-user-images.githubusercontent.com/170263/331862632-dbf77435-c665-474a-9349-2161c49c6e03.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYxMTMyNDEsIm5iZiI6MTcxNjExMjk0MSwicGF0aCI6Ii8xNzAyNjMvMzMxODYyNjMyLWRiZjc3NDM1LWM2NjUtNDc0YS05MzQ5LTIxNjFjNDljNmUwMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxOVQxMDAyMjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05NjY3Mjk4Y2UwMTVhMDgyNDVmYjc4ZGE4OGQzNTg3YTFlMzNjNzRhZmVlYzQ2MWEyZDUwNjIxNjFmMjhkZTE3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.u0bYeMQKzwpaiLXEaAc9yOdG3cyTLTWHk0SA9rZ3nlA">
> 
> It is not clickable, so there's not much sense to share it in exactly this way.

No, that's not true. If that message is send it will look like the message you see above that. It is labeled "location = Standort" and will open a map if pressed.

**Comment by licaon-kter** [comment]
@biodranik that's the INPUT FIELD of a chat app :))

You can see the button it will result above it,

**Comment by Menelmacar** [comment]
Here you can see how it transforms to a link. ( will happen if more then the geo link is in the message) 
If only geo link then it will show a button 
![Imagepipe_4](https://github.com/organicmaps/organicmaps/assets/11552677/9b4f8f0c-9af8-4276-abc2-072ebb6b4940)

Bu all of this is only one use case of geo links. 
A bug one is to share it between your own apps without any messager in the middle. As described in my first and second message.

**Comment by jech** [comment]
CC @M66B.

**Comment by M66B** [comment]
@jech why are you mentioning me?

Note that this was already handled:

https://github.com/organicmaps/organicmaps/issues/7916#issuecomment-2054191890

There are changes for that in the GitHub version of the app (not yet in the Play Store version).

**Comment by biodranik** [comment]
Good, so there is one messaging app that properly displays geo: links without slashes? How many users are there?

**Comment by jech** [comment]
@M66B, I CC'd you in case you could contribute usefully to the discussion, which you have.  Thanks for your input.

**Comment by Menelmacar** [comment]
I've got no idea, because it is a decentralized network without tracking and fdroid has no stats. Let's say a million.
I still think the main usecase is to share the link on my own device with outer apps that can consume geo: links. As you may have seen there are a lot of them. Especially OpenStreetMap apps to contribute or way tracker. 
In the end you must do with this issue what you think best for your app anyway. 
I don't see the harm in sharing geo anlongside your https link as osmand does. So users can use what they want with it. Personally I like open standards, and think they are useful in its own right. Additionally to the added functionality by sharing the location from the app to other of my apps.

**Comment by biodranik** [comment]
Let's be clear and precise:
1. There is no or very little value to share `geo:13.123,12.123` amongst other texts (e.g. address, omaps.app URL, etc.) when a location is shared in a message, because it is not clickable. Copying the whole message and then deleting other text while trying to extract only coordinates or `geo:13.123,12.123` text is cumbersome.
2. A better option could be to share a more recognized _clickable_ link.
3. Another option is to copy only `geo:13.123,12.123` string of text and somehow paste/open it in another app.
4. Another option is to copy coordinates `13.123 12.123` and paste them into another app (it already works).
5. Yet another option is to launch a system action "open this `geo:13.123,12.123` URL with an app for it", so when pressing such a button in OM you immediately open the coordinate in another app.

Let's brainstorm: What would be the simplest UX for this feature without complicating OM for millions of non-geeky users?

**Comment by Menelmacar** [comment]
Expose geo: to other apps with a share button would be my idea about that. 
Share menu reveals:

Send location
share geo:

**Comment by jech** [comment]
> Another option is to copy only geo:13.123,12.123 string of text and somehow paste/open it in another app.

That would work for me.

> Yet another option is to launch a system action "open this geo:13.123,12.123 URL with an app for it", so when pressing such a button in OM you immediately open the coordinate in another app.

That would also work for me.

> Let's brainstorm: What would be the simplest UX for this feature without complicating OM for millions of non-geeky users?

For sharing bookmarks (stars), I think an extra entry in the long-press menu would be fine — it would slightly complicate the UI, but I think that most users are willing to accept an extra entry in a contextual menu.  For sharing the current location, I have frankly no idea, since we don't want to have two distinct share buttons that do subtly different things.

**Comment by patepelo** [comment]
[Signal App doesn't have any plans to implement it soon](https://github.com/signalapp/Signal-Android/issues/9189#issuecomment-1080842850)

Can we research other messaging apps covering it and list in a table? (Users that have these apps please)

Messaging App | Status | Comments
|--------|--------|--------
WhatsApp | Not supported yet |
Telegram | Not supported yet|
WeChat|Unknown|
Line |Unknown|
Viber|Unknown|
Facebook Messenger|Unknown|
Instagram|Not supported yet|
Snapchat|Unknown|
iMessage|Unknown|
Messages (Android SMS)|Not supported yet| Android version 14
Signal | Not supported yet | No plans to implement soon
KakaoTalk|Unknown|
QQ|Unknown|
Discord|Not supported yet| At leas on MacOS version

**Comment by licaon-kter** [comment]
Conversations (and forks) on Android support it, as said above

Monal on iOS and macOS

Siskin on iOS and Beagle on macOS

**Comment by patepelo** [comment]
> /PS: _Funny/sad how the list above contains only proprietary messaging silos, that would/should/could be used as a guiding rule for this issue of an app that's open source, based on open map data, and that otherwise is proudly anti-everything those silos and their master represent._

I started the list based on a search for the most popular apps and just added your other references in the list.

**Comment by RedAuburn** [comment]
LineageOS Messaging app: ❌
ProtonMail (Android): ❌
Element (Android): 👍️

**Comment by patepelo** [comment]
related to #7103

**Comment by patepelo** [comment]
> Please try "Open in Another App" feature that basically shares geo: link.
> 
> #7105
What do you mean? That it's what happening in the background when opening with another app? How can I extract the geo uri? I'm testing on latest androit telegram beta

**Comment by Misalf-git** [comment]
> How can I extract the geo uri?

You can long-tap the button to copy `geo:...` to the clipboard.

**Comment by pastk** [comment]
- related to https://github.com/organicmaps/organicmaps/pull/1821

**Comment by pastk** [comment]
> If you're concerned that messaging apps don't all support these as links. Yes that's true, but some do.
> And for the rest there can be a fallback http URL.

So ideally when a user is sharing some location via text (messenger, email, etc.) there should be several options:
```
https://omaps.app/56.63404,47.88158/Chechevitsa
geo:56.63404,47.88158
56.63404,47.88158
```
This is supposed to work for all cases, but at the same time looks confusing and redundant as its basically the same `56.63404,47.88158` but wrapped differently.
